### PR TITLE
Fix assembler for c.li, c.andi, c.addi, c.addiw and c.lui

### PR DESCRIPTION
**Pull request by kito-cheng**
- User ISA spec 2.1 p86, dest in c.addi can be x0
- User ISA spec 2.1 p85, imm in c.li can be 0
- User ISA spec 2.1 p86, dest in c.addiw can be 0
- User ISA spec 2.1 p87, dest in c.andi can be 0
- User ISA spec 2.1 p85, imm in c.lui can accept 0x0~0x3f

### CODE REVIEW

File: gas/config/tc-riscv.c [hunk 1]
```diff
@@ -1377,9 +1377,9 @@ riscv_ip (char *str, struct riscv_cl_insn *ip, expressionS *imm_expr,
 		  if (imm_expr->X_op != O_constant
 		      || imm_expr->X_add_number <= 0
 		      || imm_expr->X_add_number >= RISCV_BIGIMM_REACH
-		      || (imm_expr->X_add_number >= RISCV_RVC_IMM_REACH / 2
```
Comment (sorear) [inline gas/config/tc-riscv.c]: Is this actually correct?  "C.LUI loads the non-zero 6-bit immediate field into bits 17–12 of the destination register, clears the bottom 12 bits, and sign-extends bit 17 into all higher bits of the destination"  Sounds like a -32..31 to me, and qemu appears to implement it that way as well.

Comment (kito-cheng) [inline gas/config/tc-riscv.c]: Here is some discussion about lui: https://github.com/riscv/riscv-binutils-gdb/issues/59

Comment (aswaterman) [inline gas/config/tc-riscv.c]: I believe this is basically the same issue as #59.  While we made a bad decision about how to expose LUI, we are now stuck with it.

**Comment by kito-cheng** [comment]
Patch updated :)

**Comment by palmer-dabbelt** [comment]
4368ca4 Fix assembler for c.li, c.andi and c.addiw
4e9d526 Fix assembler for c.addi, rd can be x0

**Pull request closed without merging**